emscripten.py

```diff
@@ -1666,6 +1666,10 @@ def create_receiving(function_table_data, function_tables_defs, exported_impleme
 
   shared.Settings.MODULE_EXPORTS = module_exports = exported_implemented_functions + function_tables(function_table_data)
 
+  # Should only need to do this in the main thread.
+  if shared.Settings.USE_PTHREADS:
+    receiving = "if (!ENVIRONMENT_IS_PTHREAD) {\n" + receiving + "}\n"
```

Comment:
But the stuff added to receiving below we do want on non-main-threads?  Can you explain, perhaps write more in the comment here?

Reply:
In my testing it seems like these functions were already shared with the non-main-threads by the main thread. Calling these from a non-main-thread attempted to redefine them (as the same thing). No idea if that's how it's supposed to work.

Reply:
Based on where this line is, it just puts the runtime assertions stuff behind this check. What's special about them? As @sbc100 said, I think this should all be needed in pthreads too - do you have a test that fails without this change?

Reply:
When I tried to run the code before this change I was getting errors about the vars in this block being redefined. It seems like when initialising a thread these functions have already been defined (not sure how). Normally var redefinition isn't an issue as everything gets hoisted, so I assumed it was an ES6 module thing. I'll have to do some more testing, I can't remember exactly what error I was trying to avoid.

Although looking at that again I should add a check for if `receiving` is an empty string.

Reply:
That's a very odd error - I don't think we use read-only properties anywhere?

Reply:
Yeah my guess is that in ES6 modules properties are read-only by default. I don't see any way to change that without changing the way we define properties. That seems like a lot of work for very little gain.
I was hoping I could just change it to writable after creation, like `Object.defineProperty(asm, '__ZSt18uncaught_exceptionv', {configurable:true, writable:true});` but that just throws `Uncaught TypeError: Cannot redefine property: __ZSt18uncaught_exceptionv`.

The affect is pretty edge case, and the impact is that there's slightly less debugging statements. Personally I don't think this is that big of a deal, do you? This could always get merged and an issue opened to fix runtime assertions inside an ES6 module.

Reply:
Actually after taking another look, the code looks like:

```
var real___ZSt18uncaught_exceptionv = asm["__ZSt18uncaught_exceptionv"];
asm["__ZSt18uncaught_exceptionv"] = function() {
  assert(runtimeInitialized, 'you need to wait for the runtime to be ready (e.g. wait for main() to be called)');
  assert(!runtimeExited, 'the runtime was exited (use NO_EXIT_RUNTIME to keep it alive after main() exits)');
  return real___ZSt18uncaught_exceptionv.apply(null, arguments);
};
```
...
```
var __ZSt18uncaught_exceptionv = Module["__ZSt18uncaught_exceptionv"] = function() {
  assert(runtimeInitialized, 'you need to wait for the runtime to be ready (e.g. wait for main() to be called)');
  assert(!runtimeExited, 'the runtime was exited (use NO_EXIT_RUNTIME to keep it alive after main() exits)');
  return asm["__ZSt18uncaught_exceptionv"].apply(null, arguments)
};
```

So if the code calls `Module.__ZSt18uncaught_exceptionv` it already goes through the runtime checks. So it seems like that entire block of code may be redundant in WASM anyway.

Reply:
I'm having a hard time following this - what is failing, and why? I don't see how to connect the error message from before about read-only properties to those code quotes.

Reply:
So it fails on the line `asm["__ZSt18uncaught_exceptionv"] = function() {`. with `Cannot redefine property: __ZSt18uncaught_exceptionv`
I'm not sure where this property is initially defined, probably in the WASM?
The function is modified so that it goes through runtime assertions before being called. However when the function `Module["__ZSt18uncaught_exceptionv"]` is called it goes through the assertions, into `asm["__ZSt18uncaught_exceptionv"]`, then through the same assertions again. It seems a little redundant to me, doubling up the assertions.

So to sum up:
**The bug:**
The line `asm["__ZSt18uncaught_exceptionv"] = function() {`. throws with `Cannot redefine property: __ZSt18uncaught_exceptionv`

**My change:**
Don't redefine (to inject runtime assertions) functions when inside an ES6 module pthread.

**The impact:**
In an ES6 module pthread asm functions will only jump through one runtime assertion check, instead of the usual two.

If it would make it clearer I could post some example code, but you would either need to build it off my branch, or I could post the compiled example.

Reply:
Yeah, maybe post the compiled example, and instructions for how to run it. I've never seen that error before, I'm mystified.

Are you perhaps running this with `use strict` or some special JS VM flag?

Reply:
I tried to attach a zip to this comment, but it didn't work. [Hopefully you can access this link?](https://hexmet-my.sharepoint.com/:u:/g/personal/tim_lander_hexagon_com/EX0XSdcY6GtDmtpM-7Eb16IBgn29uA-_ojG1F41NXzsRlw?e=vcB2WN)

It can only be run though Chrome with the "Experimental Web Platform features" flag enabled.
I ran the server like: `emrun --no_browser --no_emrun_detect --hostname $(hostname -I) --port 8080 module_test/test.html`, so nothing fancy

If you look at `line 3291` you'll see this change. If you change that line to `if (true) {` the code will throw with `TypeError: Cannot assign to read only property '__ZN14test_wrapper_212pthread_testEi' of object '[object Object]'`

Hopefully this makes things clearer.

Reply:
Thanks for the testcase - I'm not sure I understand the instructions after it, though. I set that flag, and then on line 3291 I see
```
var _setThrew = Module["_setThrew"] = function() {
```
and I changed that to
```
if (true) {
```
Now when I run it I get `you need to wait for the runtime to be ready`. So I don't see that "read only property" error?

Reply:
Uh, I think you're looking at the wrong line. `var _setThrew = Module["_setThrew"] = function() {` is on line 3**92**1, I want you to look at line 3**29**1.

Line 3291 should have `if (!ENVIRONMENT_IS_PTHREAD) {`

Reply:
Thanks, and sorry for my mistake earlier!

Ok, what I think is happening here is: when loaded as an ES6 module, the wasm exports object is an ES6 module export object, or behaves like one, and those are more "static" than normal JS objects. And we try to replace them with wrappers to add extra runtime error handling, and that fails.

But that doesn't explain why this is a problem only on workers but not on the main thread. It sounds like that could be a browser bug, in fact. That asm object is a wasm export object on both locations, so it should behave the same? Unless someone only the worker is an ES6 context - I'm not too familiar with how those are created?

Reply:
Regardless, we can split this out from the rest of this PR, if it's separable.

Reply:
Yeah I don't really know too much about all that stuff either.
The change on that line allows EXPORT_ES6 + USE_PTHREADS to work with ASSERTIONS != 0. I think the best thing is to merge this PR as is, since by default ASSERTIONS=1.

Perhaps after this is merged another bug can be opened to fix those assertions?\

If you'd prefer I'm happy to remove that line, and just say that currently EXPORT_ES6 + USE_PTHREADS only work with ASSERTIONS=0.

Reply:
Yeah, let's separate it out - a comment that says something like your last line there sounds good enough for now.

Reply:
OK, I changed it to an assert, since the next line would just throw anyway. At least this way the developer gets a clear message about this not working. Once this is merged I'll open another issue to fix USE_PTHREADS + EXPORT_ES6 +ASSERTIONS.